File: scripts/python/p6_research_context.py
"""
P6 research context loader — closed loop → evolution + cognition.
Reads data/p6_research_context.json from egx_closed_loop.mjs.
"""
from __future__ import annotations

import json
from pathlib import Path

ROOT = Path(__file__).resolve().parents[2]
CONTEXT_PATH = ROOT / "data" / "p6_research_context.json"


def load_context() -> dict | None:
    if not CONTEXT_PATH.exists():
        return None
    try:
        return json.loads(CONTEXT_PATH.read_text(encoding="utf-8"))
    except Exception:
        return None


def _ctx_from_params(params: dict | None) -> dict | None:
    if params and params.get("p6_context"):
        return params["p6_context"]
    return load_context()


def ingest_p6_ultra_failures(db, params: dict | None = None) -> dict:
    """P6 live ULTRA losses → failure_reconstruction (closed loop)."""
    ctx = _ctx_from_params(params) or {}
    downrank = set(ctx.get("evolution_hints", {}).get("downrank_behavioral") or [])

    try:
        losses = db.execute("""
            SELECT symbol, signal_date, behavioral_class, return_t5
            FROM recommendation_outcomes
            WHERE conviction_tier = 'ULTRA_CONVICTION'
              AND outcome_filled >= 5
              AND hit_t5 = 0
            ORDER BY signal_date DESC
            LIMIT 30
        """).fetchall()
    except Exception:
        return {"n_ingested": 0, "reason": "NO_RECOMMENDATION_OUTCOMES"}

    n = 0
    symbols = []
    for f in losses:
        cls = f["behavioral_class"] or "UNKNOWN"
        primary = "P6_ULTRA_LOSS"
        if cls in downrank:
            primary = f"P6_{cls}_DOWNRANK"
        try:
            db.execute("""
                INSERT INTO failure_reconstruction
                  (failure_date, symbol, law_id, law_name, direction,
                   failure_class, primary_cause, secondary_cause,
                   regime_at_failure, feature_value_at_failure,
                   n_competing_laws, reconstruction_confidence)
                VALUES (?,?,?,?,?,?,?,?,?,?,?,?)
            """, (
                f["signal_date"], f["symbol"], "P6_LIVE", "ULTRA_CONVICTION", "UP",
                primary, primary, cls,
                "", None, 0, 0.85,
            ))
            n += 1
            symbols.append(f["symbol"])
        except Exception:
            pass

    if n:
        db.commit()

    findings = []
    if n:
        findings.append(f"P6 live: {n} ULTRA losses ingested into failure_reconstruction")
    explosive_n = sum(1 for f in losses if (f["behavioral_class"] or "") == "EXPLOSIVE")
    if explosive_n and "EXPLOSIVE" in downrank:
        findings.append(f"P6 priority: EXPLOSIVE downrank ({explosive_n} recent ULTRA losses)")

    return {
        "n_ingested": n,
        "symbols": symbols[:15],
        "key_findings": findings,
        "downrank_targets": list(downrank),
    }


def apply_p6_stock_adjustments(db, params: dict | None = None) -> dict:
    """Post stock_behavioral_memory — bump false_signal_rate from P6 outcomes."""
    ctx = _ctx_from_params(params) or {}
    hints = ctx.get("evolution_hints") or {}
    downrank = set(hints.get("downrank_behavioral") or [])
    loss_symbols = set(hints.get("loss_symbols") or [])

    adjusted = []
    class_bumps = 0

    for cls in downrank:
        try:
            cur = db.execute("""
                UPDATE stock_behavioral_memory
                SET false_signal_rate = MIN(0.95, COALESCE(false_signal_rate, 0.5) + 0.08),
                    mutation_flag = 1
                WHERE behavioral_class = ?
            """, (cls,))
            class_bumps += cur.rowcount
        except Exception:
            pass

    for sym in loss_symbols:
        try:
            cur = db.execute("""
                UPDATE stock_behavioral_memory
                SET false_signal_rate = MIN(0.98, COALESCE(false_signal_rate, 0.5) + 0.10),
                    mutation_flag = 1
                WHERE symbol = ?
            """, (sym,))
            if cur.rowcount:
                adjusted.append(sym)
        except Exception:
            pass

    if adjusted or class_bumps:
        db.commit()

    findings = []
    if class_bumps:
        findings.append(f"P6 adjusted false_signal_rate for {class_bumps} {','.join(downrank)} stocks")
    if adjusted:
        findings.append(f"P6 penalized {len(adjusted)} ULTRA-loss symbols in behavioral memory")

    return {
        "class_rows_bumped": class_bumps,
        "symbol_rows_bumped": len(adjusted),
        "symbols": adjusted[:15],
        "key_findings": findings,
    }


def apply_p6_cognition_priorities(db, params: dict | None = None) -> dict:
    """Extract P6 priorities for cognition pipeline key_findings."""
    ctx = _ctx_from_params(params) or {}
    hints = ctx.get("cognition_hints") or {}
    priorities = []

    if hints.get("prioritize_explosive_review"):
        priorities.append({
            "focus": "EXPLOSIVE",
            "reason": "P6 forensic downrank — review explosion archetypes",
            "loss_count": hints.get("explosion_loss_count", 0),
        })

    for flag in hints.get("pattern_flags") or []:
        priorities.append({
            "focus": flag,
            "reason": "P6 loss autopsy pattern — investigate in self_evolution",
        })

    gate = ctx.get("p6_gate") or {}
    findings = []
    if gate.get("n_completed"):
        wr = gate.get("win_rate")
        wr_s = f"{wr:.1f}%" if wr is not None else "—"
        findings.append(
            f"P6 gate: {gate['n_completed']}/{gate.get('min_n', 30)} @ {wr_s} WR"
        )
    if priorities:
        findings.append(f"P6 cognition priorities: {len(priorities)} active")

    return {
        "priorities": priorities,
        "key_findings": findings,
        "p6_context_loaded": bool(ctx),
    }

File: scripts/lib/p6_research_context.mjs
/**
 * P6 research context — bundles closed-loop outputs for evolution/cognition/discovery.
 */
import Database from 'better-sqlite3';
import { existsSync, readFileSync, writeFileSync, mkdirSync } from 'fs';
import { join } from 'path';
import { PROJECT_ROOT } from './load_env.mjs';
import { DB_PATH } from './delivery_audit.mjs';
import { getProofLoopMetrics, PROOF_MIN_N, PROOF_MIN_WR } from './proof_loop.mjs';
import { readDiscoveryFeedback } from './discovery_feedback.mjs';

export const P6_CONTEXT_PATH = join(PROJECT_ROOT, 'data/p6_research_context.json');

function queryUltraLosses(limit = 30) {
  if (!existsSync(DB_PATH)) return [];
  const db = new Database(DB_PATH, { readonly: true });
  try {
    const hasTable = db.prepare(`
      SELECT name FROM sqlite_master WHERE type='table' AND name='recommendation_outcomes'
    `).get();
    if (!hasTable) return [];
    return db.prepare(`
      SELECT symbol, signal_date, behavioral_class, return_t5, hit_t5
      FROM recommendation_outcomes
      WHERE conviction_tier = 'ULTRA_CONVICTION'
        AND outcome_filled >= 5
        AND hit_t5 = 0
      ORDER BY signal_date DESC
      LIMIT ?
    `).all(limit);
  } finally {
    db.close();
  }
}

function queryUltraWins(limit = 15) {
  if (!existsSync(DB_PATH)) return [];
  const db = new Database(DB_PATH, { readonly: true });
  try {
    const hasTable = db.prepare(`
      SELECT name FROM sqlite_master WHERE type='table' AND name='recommendation_outcomes'
    `).get();
    if (!hasTable) return [];
    return db.prepare(`
      SELECT symbol, signal_date, behavioral_class, return_t5, hit_t5
      FROM recommendation_outcomes
      WHERE conviction_tier = 'ULTRA_CONVICTION'
        AND outcome_filled >= 5
        AND hit_t5 = 1
      ORDER BY signal_date DESC
      LIMIT ?
    `).all(limit);
  } finally {
    db.close();
  }
}

export function buildP6ResearchContext({
  signalDate = null,
  learning = null,
  forensic = null,
  discovery = null,
  opportunity = null,
  discoveryQuality = null,
  oppFollowup = null,
  ingested = null,
} = {}) {
  const proofAll = getProofLoopMetrics();
  const proofDelivered = getProofLoopMetrics({ deliveredOnly: true });
  const feedback = discovery || readDiscoveryFeedback();
  const ultraLosses = queryUltraLosses();
  const ultraWins = queryUltraWins();

  const downrank = (feedback?.queue || [])
    .filter(i => i.type === 'DOWNRANK_BEHAVIORAL')
    .map(i => i.target);
  const investigate = (feedback?.queue || [])
    .filter(i => i.type === 'INVESTIGATE_PATTERN')
    .map(i => i.target);

  const ctx = {
    at: new Date().toISOString(),
    signal_date: signalDate,
    p6_gate: {
      n_completed: proofAll.n_completed,
      win_rate: proofAll.win_rate,
      gate_pass: proofAll.gate_pass,
      samples_needed: proofAll.samples_needed,
      min_n: PROOF_MIN_N,
      min_wr: PROOF_MIN_WR,
    },
    p6_delivered: {
      n_completed: proofDelivered.n_completed,
      win_rate: proofDelivered.win_rate,
      delivered_only: true,
    },
    discovery_feedback: feedback,
    forensic_summary: forensic ? {
      n_losses: forensic.n_losses,
      by_class: forensic.by_class,
    } : null,
    loss_autopsy: learning?.loss_autopsy ? {
      n_residual_losses: learning.loss_autopsy.n_residual_losses,
      flag_counts: learning.loss_autopsy.flag_counts,
    } : null,
    counterfactual: learning?.counterfactual ? {
      projected_wr: learning.counterfactual.projected_wr,
      would_block_losses: learning.counterfactual.would_block_losses,
      would_block_wins: learning.counterfactual.would_block_wins,
    } : null,
    opportunity_snapshot: opportunity ? {
      quality_score: opportunity.quality_score,
      n_delivered: opportunity.n_delivered,
      n_top_opportunity: opportunity.n_top_opportunity,
      missed_high_opportunity: opportunity.missed_high_opportunity?.length ?? 0,
      discovery_quality_score: opportunity.discovery_quality_score ?? discoveryQuality?.discovery_quality_score,
    } : null,
    discovery_quality: discoveryQuality ? {
      score: discoveryQuality.discovery_quality_score,
      grade: discoveryQuality.grade,
      quant_avg_quality: discoveryQuality.quant?.avg_quality ?? null,
      sweet_spot_rules: discoveryQuality.quant?.sweet_spot_rules ?? null,
      opp_qualified_pct: discoveryQuality.components?.opp_qualified_pct ?? null,
      near_ath_risk: discoveryQuality.opportunity?.near_ath_risk_count ?? null,
    } : null,
    opportunity_trend: oppFollowup ? {
      alerts: oppFollowup.alerts?.length ?? 0,
      trends: oppFollowup.trends,
      top_alert: oppFollowup.alerts?.[0]?.code ?? null,
    } : null,
    ultra_losses: ultraLosses,
    ultra_wins: ultraWins,
    evolution_hints: {
      downrank_behavioral: downrank,
      investigate_patterns: investigate,
      loss_symbols: ultraLosses.map(r => r.symbol),
      win_symbols: ultraWins.map(r => r.symbol),
    },
    cognition_hints: {
      prioritize_explosive_review: downrank.includes('EXPLOSIVE'),
      explosion_loss_count: ultraLosses.filter(r => r.behavioral_class === 'EXPLOSIVE').length,
      pattern_flags: investigate,
    },
    directives_ingested: ingested?.ingested ?? 0,
    research_priorities: (feedback?.queue || []).slice(0, 8).map(q => ({
      type: q.type,
      target: q.target,
      priority: q.priority,
      rationale: q.rationale,
    })),
  };

  return ctx;
}

export function writeP6ResearchContext(ctx) {
  mkdirSync(join(PROJECT_ROOT, 'data'), { recursive: true });
  writeFileSync(P6_CONTEXT_PATH, JSON.stringify(ctx, null, 2));
  return { path: P6_CONTEXT_PATH, at: ctx.at };
}

export function loadP6ResearchContext() {
  if (!existsSync(P6_CONTEXT_PATH)) return null;
  try {
    return JSON.parse(readFileSync(P6_CONTEXT_PATH, 'utf8'));
  } catch {
    return null;
  }
}
